### PR TITLE
convert guava optionals to java

### PR DESCRIPTION
**Pull request by tpetracca**
**Goals (and why)**:
Convert all compile time usage of guava Optionals to java Optionals. Generally good cleanup, plus potentially help any consumers of ours who have to do annoying manual conversions by allowing them to remove that cruft from their codebases (I believe in practice most of our consumers use java optionals).

**Implementation Description (bullets)**:
Impl is pretty straight forward, I just used some git and bashfu to auto do the import changes, had eclipse fix the imports ordering, and then mostly manually fixed the breaks.
For the breaks that had to be manually fixed, I mostly followed guava's own recommendations about what java optional equivalents map to their optional's methods. You can fine these in [Guava's javadocs](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Optional.html).
It is important to note that I **did not touch the table rendering code**, and so `git grep "com.google.common.base.Optional"` will still return some hits in `atlasdb-client`.

**Concerns (what feedback would you like?)**:
Is this a breaking change for consumers (both products and deployments)?
For deployments? It is not. Jackson [de]serialization magic should mean that json/yml config blobs will be unchanged despite the compile time Optional type change.
For developers? Probably a break. Anyone manually building config objects using their immutable builders is likely going to have compile time breaks and will have to switch over to guava. As mentioned above, this is actually likely a good thing as it will entail deleting gross conversion code because a lot of our clients are already mostly using java Optionals.

The specific projects and classes where we may be causing a break are that might be worth paying special attention to:
- `atlasdb-api`
- `atlasdb-config`
- `lock-api`
- `timelock-server`
- `atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfig.java`
- `atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/config/OracleConnectionConfig.java`
- `leader-election-api/src/main/java/com/palantir/leader/LeaderElectionService.java`

**Where should we start reviewing?**:
Probably the classes I outlined above, plus anything that wasn't an obvious port (like the code in `AtlasDbConfigs.java`.

**Priority (whenever / two weeks / yesterday)**:
Not high, but this is something that because of how widespread it is will continually get merge conflicts as develop moves forward. 

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2016)

### CODE REVIEW

**Comment by tpetracca** [comment]
@berler @gbonik @dxiao could one of you also review this from the perspective of your product. As mentioned in the notes I expect this to be a break for developer, but not existing customers, and want to make sure that is indeed true for you guys.

**Comment by tpetracca** [comment]
Tests are failing here because of error-prone being unhappy with some of the Immutable* files that are generated from classes with Optional arrays. For example the diff in `SweepResults`
```
package com.palantir.atlasdb.keyvalue.api;

-import org.immutables.value.Value;
+import java.util.Optional;

-import com.google.common.base.Optional;
+import org.immutables.value.Value;

 @Value.Immutable
```
Results in the following *partial* diff in `ImmutableSweepResults`
```
 public final class ImmutableSweepResults extends SweepResults {
-  private final Optional<byte[]> previousStartRow;
-  private final Optional<byte[]> nextStartRow;
+  private final @Nullable byte[] previousStartRow;
+  private final @Nullable byte[] nextStartRow;
   private final long cellTsPairsExamined;
```
and
```
   private boolean equalTo(ImmutableSweepResults another) {
-    return previousStartRow.equals(another.previousStartRow)
-        && nextStartRow.equals(another.nextStartRow)
+    return Objects.equal(previousStartRow, another.previousStartRow)
+        && Objects.equal(nextStartRow, another.nextStartRow)
         && cellTsPairsExamined == another.cellTsPairsExamined
```
which results in the following sadness in error-prone
```
/Volumes/git/code/atlasdb/atlasdb-api/build/classes/main/com/palantir/atlasdb/keyvalue/api/ImmutableSweepResults.java:206: error: [ArrayEquals] Reference equality used to compare arrays
    return Objects.equal(previousStartRow, another.previousStartRow)
                        ^
    (see http://errorprone.info/bugpattern/ArrayEquals)
  Did you mean 'return Arrays.equals(previousStartRow, another.previousStartRow)'?
/Volumes/git/code/atlasdb/atlasdb-api/build/classes/main/com/palantir/atlasdb/keyvalue/api/ImmutableSweepResults.java:207: error: [ArrayEquals] Reference equality used to compare arrays
        && Objects.equal(nextStartRow, another.nextStartRow)
                        ^
    (see http://errorprone.info/bugpattern/ArrayEquals)
  Did you mean '&& Arrays.equals(nextStartRow, another.nextStartRow)'?
/Volumes/git/code/atlasdb/atlasdb-api/build/classes/main/com/palantir/atlasdb/keyvalue/api/ImmutableSweepResults.java:220: error: [ArrayHashCode] hashcode method on array does not hash array contents
    h += (h << 5) + Objects.hashCode(previousStartRow);
                                    ^
    (see http://errorprone.info/bugpattern/ArrayHashCode)
  Did you mean 'h += (h << 5) + Arrays.hashCode(previousStartRow);'?
/Volumes/git/code/atlasdb/atlasdb-api/build/classes/main/com/palantir/atlasdb/keyvalue/api/ImmutableSweepResults.java:221: error: [ArrayHashCode] hashcode method on array does not hash array contents
    h += (h << 5) + Objects.hashCode(nextStartRow);
                                    ^
    (see http://errorprone.info/bugpattern/ArrayHashCode)
  Did you mean 'h += (h << 5) + Arrays.hashCode(nextStartRow);'?
4 errors
```

The funny thing about all this as that even the previous version of the `ImmutableSweepResults` is actually bad/incorrect, error-prone just doesn't catch it because it doesn't do validation of the `T` in guava `Optional<T>`, but it gets caught when using java Optional because the rendered code is a `@Nullable T` and not an `Optional<T>`.

So either way our generated code has some bad stuff in it, even before this PR goes in. Going to figure out if there's a way to work around this, but the immutables folks apparently have decided using `Optional<Array>` is not a thing they support. @bavardage ran into this at some point as well, see https://github.com/immutables/immutables/issues/554

**Comment by tpetracca** [comment]
This is blocked on #2020

**Comment by tpetracca** [comment]
Okay, so I decided to work around the issue by setting a `@Default.Value` for the `Optional<bye[]>` types in `SweepResults`, thereby forcing Immutables to directly use an `Optional<byte[]>` in `ImmutableSweepResults` (which it was previously doing when just being a guava optional). The result is that there is now no change in `ImmutableSweepResults` despite the change in `SweepResults`, and thus it isn't caught by error-prone and now my builds pass.

The problem of the bad generated code is still very real today, but is no better or worse than it was before this PR. We can trick fixing that in #2020.

**This PR should now be ready for review.** Sorry for all the traffic here, before the tests actually passed!

**Comment by jeremyk-91** [comment]
#1999 and #2006 have merged, so I'm going to update this to develop and proceed from there.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java [hunk 1]
```diff
@@ -179,8 +179,8 @@ private SweepResults doRun(TableReference tableRef,
                 lastRow = batch.lastCellExamined().getRowName();
             }
             return SweepResults.builder()
-                    .previousStartRow(startRow)
-                    .nextStartRow(Arrays.equals(startRow, lastRow) ? Optional.absent() : Optional.of(lastRow))
+                    .previousStartRow(Optional.of(startRow))
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java]: Searched for `previousStartRow` internally and it doesn't seem like anyone else directly manipulates `SweepResults` (as you might expect), so this change looks safe.

**Comment by gmaretic** [comment]
@jeremyk-91 LGTM

**Pull request merged**